### PR TITLE
sessionctx/variable: add system variable tidb_opt_range_max_size

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: ref #37176

Problem Summary:

### What is changed and how it works?

Add system variable `tidb_opt_range_mem_quota`. Part1 of #37160.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -1037,6 +1041,7 @@ const (
 	MinExpensiveQueryTimeThreshold           uint64 = 10 // 10s
 	DefTiDBConstraintCheckInPlacePessimistic        = true
 	DefTiDBForeignKeyChecks                         = false
+	DefTiDBOptRangeMemQuota                         = 67108864 // 64 MB
```
Comment (hawkingrei) [inline sessionctx/variable/tidb_vars.go:1044]: You can try to use ```util/size.MB``` to make the code more readable.

**Review by tiancaiamao — APPROVED** [review]

**Review by Yisaer — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

File: sessionctx/variable/tidb_vars.go [hunk 2]
```diff
@@ -810,6 +811,10 @@ const (
 	TiDBDDLEnableFastReorg = "tidb_ddl_enable_fast_reorg"
 	// TiDBDDLDiskQuota used to set disk quota for lightning add index.
 	TiDBDDLDiskQuota = "tidb_ddl_disk_quota"
+	// TiDBOptRangeMaxSize is the max memory limit for ranges. When the optimizer estimates that the memory usage of complete
+	// ranges would exceed the limit, it chooses less accurate ranges such as full range. 0 indicates that there is no memory
+	// limit for ranges.
+	TiDBOptRangeMaxSize = "tidb_opt_range_max_size"
```
Comment (Reminiscent) [inline sessionctx/variable/tidb_vars.go:817]: It seems wrong here. You can see line747 `// TiDB vars that have only global scope`.

Comment (xuyifangreeneyes) [reply sessionctx/variable/tidb_vars.go:817]: OK. I will put it in the right place.

Comment (xuyifangreeneyes) [reply sessionctx/variable/tidb_vars.go:817]: #37835 fixes the problem.

**Pull request merged**